### DISCUSSION TITLE
Parcel 2 transformers "asset.addIncludedFile()"

### DISCUSSION THREAD

**Discussion by itsdouges** [parcel-bundler/parcel#5677] [Q&A] [answered]
Hello! I'm trying to understand the use of `asset.addIncludedFile()` in a transformer. My assumption was that its used when one file is "brought in" to another file, essentially compiled away, it allows when modification of that file that is "brought in" to recompiled its parent.

E.g:

```
// a.js
import helloWorld from './b';

console.log(helloWorld);
```

```
// b.js
export default 'hello world';
```

When bundled looks like:

```
// a.js
console.log('hello world');
```

When changing `b.js` it would recompile `a.js`.

Is my assumption correct? Because this isn't working for me. You can see my spike here https://github.com/atlassian-labs/compiled/compare/parcel-plugin?expand=1#diff-ccae85a408034390e25030338c721b54b9c402ab058c538962a28ba26115488c

**Comment by mischnic** [comment, marked as answer]
Yes, by calling `asset.addIncludedFile(x)`, `asset` is retransformed after `x` was modified. And that seems to work fine in isolation.

So in your case, modifying `parcel/src/module.js` should retrigger the `console.log` statement. Maybe the progressbar is overwriting your log output? You could try doing `yarn start:parcel | cat` to the invocation to disable the line-rewriting spinner.

I tried cloning your repo with the parcel branch and running `yarn start:parcel` but I get a build error.

**Reply by itsdouges** [reply]
Thanks for helping @mischnic! lovely that's all I need to know - that `addIncludedFile` does what I thought it would. The branch I linked has hardcoded hacks to just work on my machine, so no worries there - I think our caching will be affecting the run as well, probably.

**Reply by itsdouges** [reply]
I've made a change but I still can't seem to get it working. Any ideas?

1. Clone https://github.com/atlassian-labs/compiled/pull/512
2. `yarn`
3. `yarn start:parcel`
4. See blue text 
5. Modify https://github.com/atlassian-labs/compiled/blob/0e938b237e3d93b33ab79905600c643d08685524/examples/packages/parcel/src/module.js to be pink or anything
6. See blue text, still. Should be updated but it isn't

I've turned caching off. Not sure if I'm missing anything? 🤔

**Reply by mischnic** [reply]
> I've made a change

I don't see a new commit after my last comment and there are still hardcoded paths?

**Reply by itsdouges** [reply]
Ooops! Try again

```
➜  compiled git:(parcel-plugin) git push
Enumerating objects: 29, done.
Counting objects: 100% (29/29), done.
Delta compression using up to 12 threads
Compressing objects: 100% (13/13), done.
Writing objects: 100% (15/15), 1.18 KiB | 1.18 MiB/s, done.
Total 15 (delta 11), reused 0 (delta 0)
remote: Resolving deltas: 100% (11/11), completed with 11 local objects.
To github.com:atlassian-labs/compiled.git
   f38333b..0e938b2  parcel-plugin -> parcel-plugin
```

**Reply by itsdouges** [reply]
I might have to simplify this to just not use Compiled to make sure it works - but currently it's not recompiling `app.js` when `module.js` changes 🤔

**Reply by wbinnssmith** [reply]
This[0] is also a relative path (or module specifier), not an absolute path. It's up to your transformer to resolve the file. We do pass a resolve function to `transform()`: 

https://github.com/parcel-bundler/parcel/blob/ac6a0cbc96ce63a43f38a6ec6c34d4c2943010a8/packages/core/types/index.js#L771

https://github.com/atlassian-labs/compiled/blob/0e938b237e3d93b33ab79905600c643d08685524/packages/parcel-transformer/src/index.tsx#L24

**Reply by itsdouges** [reply]
@wbinnssmith so i should call resolve with the relative path? let me try

**Reply by itsdouges** [reply]
atm i'm not seeing `addIncludedFile` do the behaviour i want - which is re-running the transform on `module.js` and then `app.js`

i added some more logging - only `module.js` is being re-transformed after

i see a few internal parcel transformers are using this API, unsure what I'm missing 🤔 https://github.com/parcel-bundler/parcel/blob/092457a07031d4a257a9538b4d325c9be9569c70/packages/transformers/pug/src/PugTransformer.js#L52

**Comment by itsdouges** [comment]
Hi all - still tinkering away at this https://github.com/atlassian-labs/compiled/pull/512 - It's not hard coded anymore and I've implemented `parse` and `generate` as suggested by @wbinnssmith. 

Unfortunately it's still not working - the included files aren't causing a recompilation of the files that are including them. Can anyone take a look?

@wbinnssmith @mischnic @devongovett 🤔 

1. Clone atlassian-labs/compiled#512
2. yarn
3. yarn start:parcel
4. Modify https://github.com/atlassian-labs/compiled/blob/0e938b237e3d93b33ab79905600c643d08685524/examples/packages/parcel/src/module.js to be pink or anything
5. Would expect app.js to be recompiled - it isn't

---

As a side HMR isn't working for me. Could it be because its in a monorepo? Using Firefox.

```
Uncaught TypeError: global is undefined
```

![image](https://user-images.githubusercontent.com/6801309/109366417-7e7f9100-78e7-11eb-8579-35b643a8e121.png)

**Reply by mischnic** [reply]
I've updated to the latest Parcel nightly and `addIncludedFile` seems to work correctly then

**Reply by itsdouges** [reply]
Oh 😅

Thanks! I'll give it a go.

**Reply by itsdouges** [reply]
I installed the latest nightly but get this error

```
Error: Failed to install @parcel/transformer-posthtml: Yarn failed to install modules
```

Using `^2.0.0-nightly.610` with the following deps:

- `parcel`
- `@parcel/plugin`

https://github.com/atlassian-labs/compiled/pull/512/commits/623497c905308443102428283daab8debbd1c719

Trying to see what the logs are printing but `verbose` option isn't doing anything. Does https://github.com/parcel-bundler/parcel/pull/1834 still exist in 2.0?

**Reply by itsdouges** [reply]
Looks like a mono repo problem

```
➜  compiled git:(parcel-plugin) yarn start:parcel --log-level verbose >> test.txt
warning package.json: No license field
@parcel/package-manager: yarn: Running this command will add the dependency to the workspace root rather than the workspace itself, which might not be what you want - if you really meant it, make it explicit by running this command again with the -W flag (or --ignore-workspace-root-check).
@parcel/package-manager: yarn: Running this command will add the dependency to the workspace root rather than the workspace itself, which might not be what you want - if you really meant it, make it explicit by running this command again with the -W flag (or --ignore-workspace-root-check).
Error: Failed to install @parcel/transformer-posthtml: Yarn failed to install modules
Error: Failed to install @parcel/transformer-posthtml: Yarn failed to install modules
    at install (/Users/madou/projects/compiled/node_modules/@parcel/package-manager/lib/installPackage.js:161:11)
    at async PromiseQueue._runFn (/Users/madou/projects/compiled/node_modules/@parcel/utils/lib/PromiseQueue.js:108:7)
    at async PromiseQueue._next (/Users/madou/projects/compiled/node_modules/@parcel/utils/lib/PromiseQueue.js:95:5)
```

**Reply by itsdouges** [reply]
Adding deps and it works a treat https://github.com/atlassian-labs/compiled/pull/512/commits/35662b0174b789e07fe5f9d43afe998089df568f

However HMR is still not working. I get a different error now though!

![image](https://user-images.githubusercontent.com/6801309/109371388-cb1f9800-78f8-11eb-83ee-4625850da981.png)

```
Uncaught TypeError: asset.depsByBundle is undefined
```

Edit: It's a string? Hmm

![image](https://user-images.githubusercontent.com/6801309/109373423-bb0cb600-7902-11eb-8f38-a387b43c1c96.png)

@mischnic any ideas 😄 thanks for the help so far.

Another edit: I'm also seeing odd behaviour after starting up again with a `.parcel-cache` already existing. The parent files that import an included file aren't being re-compiled. 

Steps to reproduce:


1. Clone atlassian-labs/compiled#512
2. yarn
3. yarn start:parcel
4. Modify examples/packages/parcel/src/module.js
5. Reload - should see the app updated (HMR not working)
6. Kill the server.
7. yarn start:parcel
8. Modify examples/packages/parcel/src/module.js
9. Reload the app
10. Notice that its not updated as expected

When running with `--no-cache` it works as expected.

**Reply by mischnic** [reply]
> asset.depsByBundle is undefined

That's a bug. The browser-side HMR runtime is a bit sketchy at the moment.

> The parent files that import an included file aren't being re-compiled.

Seems to be working for me

**Reply by itsdouges** [reply]
Let me take a video for you to illustrate what I'm seeing to make sure we're on the same page 😄. I've uploaded a video going over what I'm seeing:

https://user-images.githubusercontent.com/6801309/109402418-cff05480-79a9-11eb-8e56-d6a3fc9056d7.mov

**EXTRA NOTE 👉👉 This behaviour doesn't happen when parcel is ran with `--no-cache`.**

Since the original question has been answered - I'll make an issue for us to talk through it there.

**Reply by itsdouges** [reply]
Raised here https://github.com/parcel-bundler/parcel/issues/5927
